### ISSUE TITLE
[RFC] Add Support for PHP File Configuration

### ISSUE THREAD

**Issue by loic425** [Behat/Behat#1539]
### Description
Currently, Behat supports configuration through YAML files. While YAML is straightforward for many use cases, offering support for PHP configuration files could bring added flexibility and advantages, such as:

- Enhanced readability for developers familiar with PHP syntax.
- Leveraging PHP logic (e.g., conditions, constants, or reusable configurations).
- Improved IDE support and type checking.

### Proposal
Introduce support for configuring Behat using PHP files alongside the existing YAML configuration. This would involve:

- Parsing configurations from a behat.php file if present.
- Maintaining backward compatibility with YAML configuration files (behat.yml, behat.yaml).
-  Providing a basic example of a PHP configuration file in the documentation.

### Benefits

- Greater flexibility in defining configurations.
-  Ability to dynamically generate configurations based on environment or context.
-  Seamless integration with PHP constants, variables, and classes.

### Example

Here’s an example of what a PHP configuration file (behat.php) could look like:

```php
use Behat\Config\Config;

return new Config([
    'default' => [
        'gherkin' => [
            'filters' => [
                'tags' => '~@php8'
            ],
        ],
    ],
]);
```

### Implementation Notes

- Extend the Behat configuration loader to check for behat.php.
- Update the documentation to explain the new option and its use cases.

### Future helpers

We can do everything with a simple array. But to improve the DX, we can add some helpers to the Config object.

**Importing Configuration Files**

To make configurations modular and reusable, the PHP file should support importing other files. 

It's already possible with the previous proposal:

```php
use Behat\Config\Config;

return new Config([
    'imports' => [
        __DIR__.'/other-config.php'
    ],
])
```

But we need some helpers to improve the DX.

For instance:

```php
use Behat\Config\Config;

$config = new Config();
$config->import('config/suites.php');

return $config;
```

We can import several configuration files with the same "import" method.

```php
use Behat\Config\Config;

$config = new Config();
$config->import([
    'config/extensions.php', 
    'config/suites.php',
]);

return $config;
```

We can also introduce more smart imports.

```php
use Behat\Config\Config;

$config = new Config();
$config->import('config/**/**.php');

return $config;
```

**Adding suites**

Here's is just a proposal, it needs to be discussed.

```php
use App\Tests\Behat\Context\FirstContext;
use App\Tests\Behat\Context\SecondContext;
use Behat\Config\Config;
use Behat\Config\Profile;
use Behat\Config\Suite;

$defaultProfile = (new Profile('default'))
    ->withSuite($adminDashboardSuite);
;

$defaultProfile->withSuite(
    new Suite(
        name: 'admin_dashboard', 
        settings: [
            'contexts' => [
                FirstContext::class,
                SecondContext::class,
            ],
            'filters' => [
                'tags' => '@admin&&@dashboard'
            ],
        ],
    )
);

return new (Config())
    ->withProfile($defaultProfile)
;
```

or using more helpers

```php
use App\Tests\Behat\Context\FirstContext;
use App\Tests\Behat\Context\SecondContext;
use Behat\Config\Config;
use Behat\Config\Profile;
use Behat\Config\Suite;

$adminDashboardSuite = (new Suite(name: 'admin_dashboard'))
    ->withFilters(['tags' => ['@admin&&@dashboard']])
    ->withContexts(
        FirstContext::class, 
        SecondContext::class
    )
;

$defaultProfile = (new Profile('default'))
    ->withSuite($adminDashboardSuite)
;

return (new Config())
    ->withProfile($defaultProfile)
;
```

### Overall solution

**The main config file**

```php
use Behat\Config\Config;
use Behat\Config\Extension;
use Behat\Config\Profile;
use FriendsOfBehat\MinkDebugExtension;
use Robertfausk\Behat\PantherExtension;

$defaultProfile = (new profile('default'))
    ->import('config/suites/**/**.php')
    ->withFormatters([
        'pretty' => [
            'verbose' => true,
            'paths' => false,
            'snippets' => false,
        ],
    ])
    ->withExtension(new Extension(PantherExtension::class))
    ->withExtension(new Extension(MinkDebugExtension::class, [
        'directory' => 'etc/build',
        'clean_start' => true,
        'screenshot' => true,
    ]))
;

return (new Config())->withProfile($defaultProfile);
```

**A first suite config file**

```php
// config/suites/admin_dashboard.php
use App\Tests\Behat\Context\FirstContext;
use App\Tests\Behat\Context\SecondContext;
use Behat\Config\Config;
use Behat\Config\Profile;
use Behat\Config\Suite;

use App\Tests\Behat\Context\FirstContext;
use App\Tests\Behat\Context\SecondContext;
use Behat\Config\Config;
use Behat\Config\Profile;
use Behat\Config\Suite;

$suite = (new Suite(name: 'admin_dashboard'))
    ->withFilters(['tags' => ['@admin&&@dashboard']])
    ->withContexts(
        FirstContext::class,
        SecondContext::class,
    )
;

return (new Config())
    ->withProfile((new Profile('default'))
        ->withSuite($suite)
    )
;
```

Maybe that would be better like this. It could be a shortcut to add suites.

```php
use App\Tests\Behat\Context\FirstContext;
use App\Tests\Behat\Context\SecondContext;
use Behat\Config\Config;
use Behat\Config\Suite;

$suite = (new Suite(name: 'admin_dashboard'))
    ->withFilters(['tags' => ['@admin&&@dashboard']])
    ->withContexts(
        FirstContext::class,
        SecondContext::class,
    )
;

$config = (new Config())
    ->withSuite(suite: $suite, profile: 'default')
;
```

**Comment by acoulton** [comment]
Ahhh so the basic `Config` object just wraps a plain array but then we can add simple typehinted/self-documenting sugar methods/classes to help populate the expected schema?

So by the time the array gets into behat itself there's no difference to whether it came from yaml or PHP?

And in fact then an extension could define a DTO for its own config that people could use if they wanted to, so that helps with the extensibility question.

That could work, and probably solves my concern about maintenance. I guess we'd just be able to test the helpers by asserting that the php config files parse identically to the yaml equivalent, rather than having to run a full Behat with each type of config. That should be quite an easy way to get confidence.

I can definitely see the value in being able to reference things like class names, `E_ALL` (which people currently have to put in yaml as an int with a comment beside)

And I can see that for some people - probably including me - accessing environment variables from the PHP direct rather than learning where/how they can be used in the yaml syntax would be good.

If the php config became more commonly used in future, it could also allow us to deprecate things like some of the "magic" for guessing extension class names based on a partial name. I think that's really only there to make the yaml less verbose, but it causes a bit of confusion about the rules and how to make it work with autoloaders. For php you could just reference `MyExtension::class` and add a use statement if you wanted, which would be much simpler. Having those as clickable class references in an IDE would also make it much easier for a user to understand how the extension is wired etc.

From your fuller proposal and after thinking a bit more I am in favour of this but would like to hear what @carlos-granados and @ttomdewit think.

**Comment by carlos-granados** [comment]
I am fully in favour of this. I try to use php configuration where possible, this brings a lot of benefits, including better IDE support and the capability to lint your files using static analysers. See this blog post for more info about using this for Symfony config https://tomasvotruba.com/blog/2020/07/16/10-cool-features-you-get-after-switching-from-yaml-to-php-configs/ 
It would be great if we could get some of these benefits in Behat as well

**Comment by dragoonis** [comment]
@ciaranmcnulty :-)

**Comment by acoulton** [comment]
## Supporting DTOs etc for partial configuration

Picking up on https://github.com/Behat/Behat/pull/1537#discussion_r1853694976 and the `GenericSuite` example above, is probably a case for supporting simple DTOs that contain partial configuration. For example, as an extension writer I might want to provide a DTO with just the valid options for my extension (rather than re-implementing / extending the sugar features on `Behat\Config\Config`).

But these DTOs need to be converted back to plain arrays somewhere. There are probably two ways to do that:
* Support that on specific individual sugar methods e.g. in the `->withSuite(new GenericSuite(...))` example above, the `withSuite` method would cast to an array before adding it to `$this->settings`. This might mean either `withSuite` or `GenericSuite` having to check for and serialise nested DTOs within their arguments.
* Have the config reader take care of recursively casting objects to arrays, based on an interface.

Handling this separately in the sugar methods feels like it will get quite repetitive, and might mean adding sugar methods we don't really need just to be able to pass them an object to turn into an array.

So I think it probably makes more sense if we go for the second approach and handle this in one place when we finally transform the overall config structure back into an array in `ConfigurationLoader`.

**Comment by loic425** [comment]
@acoulton I've updated the issue to have an idea of the overall feature. I'm using every working day for ten years, I want something great to use too.

**Comment by acoulton** [comment]
@loic425 thanks for the updates on this and the PR, just to say I will make time on Saturday evening to look at things properly.

**Comment by acoulton** [comment]
@loic425 thanks for all your thought & work on this. Asking for more examples & more time was as much just that I was struggling to see the whole picture from the pieces we had so far, partly because it's taken me a while to be back to 100% brainpower after being ill. I hope I didn't come across as too critical, like you I just want to make sure we end up with something that works for everyone.

I've had a good look at this now, and I've tried converting a couple of `.yaml` from our own projects to the syntax you've proposed to get a better sense of how it looks for a bigger / more complex set of files, and how easy it is to structure & format them in the IDE. I did also experiment with using more named constructor params instead of helper methods, and actually I don't think that really helps readability/DX and is definitely less flexible.

I think the direction you're going (with my small comments on #1547 about naming / possibly skipping some of the DTO interfaces) is going to work well.

In isolation, I like your last suggestion above:

```php
$config = (new Config())
    ->withSuite(suite: $suite, profile: 'default')
;
```

But when I tried adding tester options, formatters, extensions, etc in that pattern with multiple profiles I found it quickly got hard to follow which things were in which profile. 

So I would stick with the approach you discussed with @carlos-granados and implemented in #1547, with the `Profile` object as the container for all the config helpers for a single profile.

**Comment by loic425** [comment]
@acoulton thx a lot for all your work testing the solution we're trying to cook. Don't worry, I know this kind of PRs has a lot of "ping pong".

**Comment by carlos-granados** [comment]
Looks like PHPStan is going in this direction too 😊 https://github.com/phpstan/phpstan/issues/12158

**Comment by loic425** [comment]
@acoulton @carlos-granados FYI, I still have some PRs todo for the PHP config. I'll continue tomorrow.

**Comment by carlos-granados** [comment]
> @acoulton @carlos-granados FYI, I still have some PRs todo for the PHP config. I'll continue tomorrow.

Thanks. We should also start thinking about adding some changes to our documentation

**Comment by loic425** [comment]
@carlos-granados @acoulton 
FYI, I've started a PR about PHP config documentation.
https://github.com/Behat/docs/pull/179

**Comment by carlos-granados** [comment]
@loic425 do you have plans for adding any more changes to the php config? If not, @acoulton we may have to think about doing a new release now that this and the conversion of snippets to attributes are done

**Comment by loic425** [comment]
@carlos-granados For me, it's ready for the PHP config for a new release.

I plan the "glob" support for importing file feature, but it's another thing and I don't know where I'll able to work on this.

**Comment by acoulton** [comment]
@carlos-granados @loic425 great! I think it's fine for the glob support to be an enhancement in a future release.

Definitely we should cut another release, there's just a couple of small issues that @stof spotted on the attributes PR #1549 after it was merged that we should fix first.

**Comment by carlos-granados** [comment]
I think that after merging https://github.com/Behat/Behat/pull/1569 we can close this issue. Do you agree @acoulton and @loic425 ?

**Comment by loic425** [comment]
> I think that after merging #1569 we can close this issue. Do you agree @acoulton and @loic425 ?

Yes!

**Comment by acoulton** [comment]
The only thing is I'm not sure if the docs are complete? If not, if we close this we should add an issue to Behat/Docs (and ideally tackle that soon, while the code is fresh).

**Comment by carlos-granados** [comment]
@loic425 to finish this work we need to do more work on the documentation. I have created a new issue in the `Behat/docs` repo https://github.com/Behat/docs/issues/184 and will close this one
